Python/Product/Analysis/EncodedLocation.cs

```diff
@@ -39,16 +39,16 @@ struct EncodedLocation : IEquatable<EncodedLocation> {
         public readonly object Location;
 
         public EncodedLocation(ILocationResolver resolver, object location) {
+            if (resolver == null && !(location is LocationInfo)) {
+                throw new ArgumentNullException(nameof(resolver));
+            }
+
             Resolver = resolver;
             Location = location;
         }
 
         public override int GetHashCode() {
-            if (Location != null) {
-                return Resolver.GetHashCode() ^ Location.GetHashCode();
-            }
-
-            return Resolver.GetHashCode();
+            return (Resolver?.GetHashCode() ?? 0) ^ (Location?.GetHashCode() ?? 0);
```

Comment:
FWIW, the Analysis library could really benefit from tuples in more conventional ways, seeing how often it uses `out` for multiple return values (that are also quite often discarded).

Either way, up to you.